src/test/java/htsjdk/samtools/CRAMSliceMD5Test.java

```diff
@@ -0,0 +1,105 @@
+package htsjdk.samtools;
+
+import htsjdk.samtools.cram.CRAMException;
+import htsjdk.samtools.cram.build.CramIO;
+import htsjdk.samtools.cram.ref.CRAMReferenceSource;
+import htsjdk.samtools.cram.ref.ReferenceSource;
+import htsjdk.samtools.cram.structure.Container;
+import htsjdk.samtools.cram.structure.ContainerIO;
+import htsjdk.samtools.cram.structure.CramHeader;
+import htsjdk.samtools.cram.structure.Slice;
+import htsjdk.samtools.reference.InMemoryReferenceSequenceFile;
+import htsjdk.samtools.util.SequenceUtil;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * Created by vadim on 03/07/2017.
+ */
+public class CRAMSliceMD5Test {
+
+    @Test
+    public void testSliceMD5() throws IOException {
+        final byte[] referenceBases = SequenceUtil.IUPAC_CODES_STRING.getBytes();
+        final byte[] referenceBasesUpperCased = SequenceUtil.upperCase(Arrays.copyOf(referenceBases, referenceBases.length));
+
+        final SAMFileHeader samFileHeader = new SAMFileHeader();
+        samFileHeader.addSequence(new SAMSequenceRecord("1", referenceBases.length));
+        samFileHeader.addReadGroup(new SAMReadGroupRecord("rg1"));
+
+        // this source does not change ref bases:
+        final CRAMReferenceSource referenceSourceMixedCase = (sequenceRecord, tryNameVariants) -> referenceBases;
+
+        final InMemoryReferenceSequenceFile memoryReferenceSequenceFile = new InMemoryReferenceSequenceFile();
+        // copy ref bases to avoid the original from upper casing:
+        memoryReferenceSequenceFile.add("1", Arrays.copyOf(referenceBases, referenceBases.length));
+        // this is the correct reference source, it upper cases ref bases:
+        final ReferenceSource referenceSourceUpperCased = new ReferenceSource(memoryReferenceSequenceFile);
+
+        byte[] refBasesFromUCSource = referenceSourceUpperCased.getReferenceBases(samFileHeader.getSequence(0), true);
+        // check the ref bases from the source are upper cased indeed:
+        Assert.assertEquals(refBasesFromUCSource, referenceBasesUpperCased);
+        // check there is no lower case A:
```

Comment:
This is deliberate because byte arrays are mutable. Basically this checks that referenceBasesUpperCased hasn't been crippled while normalizing bases somehow.